### PR TITLE
patch: deepcopy figure fix

### LINKED ISSUE #4925: Invalid value of type 'builtins.dict'  error when copying figure in 6.0.0rc0
The example below throws an error in 6.0.0rc0 but not in 5.24.1

```
    raise ValueError(
ValueError: 
    Invalid value of type 'builtins.dict' received for the 'x' property of scatter
        Received value: {'dtype': 'i2', 'bdata': 'oAelB6oHrwe0B7kHvgfDB8gHzQfSB9cH'}

    The 'x' property is an array that may be specified as a tuple,
    list, numpy array, or pandas Series
```

```
from dash import Dash, dcc, html
import plotly.express as px

import copy

app = Dash(__name__)

gapminder = px.data.gapminder().query("(country=='Canada') | (country=='United States') | (country=='Mexico')")

fig = px.line(gapminder, x="year", y="gdpPercap", color='country')
fig.update_yaxes(title_text='Year')
fig.update_xaxes(title_text='Adoption Rate')

# Default dimensions
fig.update_layout(
    margin=dict(l=60, r=0, t=0, b=60),
)

fig_custom_height = copy.deepcopy(fig)
fig_custom_height.update_layout(height=800)

fig_custom_short_height = copy.deepcopy(fig)
fig_custom_short_height.update_layout(height=200)

app.layout = [
                dcc.Graph(
                figure = fig_custom_height,
                ),
                dcc.Graph(
                figure = fig_custom_short_height,
            ),

]

if __name__ == '__main__':
    app.run_server(debug=True)

```

### PR DESCRIPTION
**Pull request by FBruzzesi**
## Code PR

Closes #4925 by reverting the changes in  https://github.com/plotly/plotly.py/commit/960adb9b9a89387d05343497de1df5d3df592698 to check for `is_none_or_typed_array_spec`.

Looking for indication if worth adding a test. I see that in few tests a deepcopy of figure is called, thus... I am a bit confused 🙃

### CODE REVIEW

File: packages/python/plotly/_plotly_utils/basevalidators.py [hunk 1]
```diff
@@ -1708,16 +1710,17 @@ def description(self):
...
         """.format(
             plotly_name=self.plotly_name,
-            array_ok=", or a list, numpy array or other iterable thereof"
-            if self.array_ok
-            else "",
+            array_ok=(
+                ", or a list, numpy array or other iterable thereof"
+                if self.array_ok
+                else ""
+            ),
```
Comment (FBruzzesi) [inline packages/python/plotly/_plotly_utils/basevalidators.py:1717]: 🤔 different black version?

**Comment by marthacryan** [comment]
I'd love to see a test that _fails_ before these changes! I'm a little concerned that the existing ones didn't catch this

**Comment by FBruzzesi** [comment]
Hey @marthacryan , thanks for getting back to me. Should the example in the issue enough as a test?

**Comment by stefanv** [comment]
Here's a reproducing example that came out of the skimage CI:

```python
import numpy as np

import plotly
import plotly.express as px
import plotly.graph_objects as go

plotly.io.renderers.default = "sphinx_gallery_png"

img = np.zeros((50, 50))
fig = px.imshow(img, binary_string=True)

fig.add_trace(
    go.Scatter(
        x=np.array([[1,2,3],[4,5,6]]),
        y=np.array([[1,2,3],[4,5,6]]),
        name="1"
    )
)

plotly.io.show(fig)
```

Which results in:

```
ValueError:
    Invalid value of type 'builtins.dict' received for the 'x' property of scatter
        Received value: {'dtype': 'i1', 'bdata': 'AQIDBAUG', 'shape': '2, 3'}

    The 'x' property is an array that may be specified as a tuple,
    list, numpy array, or pandas Series
```

If you toggle `binary_string` to False, it brings up another unexpected error:

```
ValueError:
    Invalid value of type 'builtins.dict' received for the 'z' property of heatmap
        Received value: {'dtype': 'f8', 'bdata': 'AAAAAAAAAAAAAAAAAA...AAA=', 'shape': '50, 50'}

    The 'z' property is an array that may be specified as a tuple,
    list, numpy array, or pandas Series
```

This is plotly 6.0.0rc0 and kaleido 1.0.0rc0.

**Comment by FBruzzesi** [comment]
I am not particularly confident with this part of the code. Minimal example failing in master:

```py
import copy
import plotly.express as px

gapminder = px.data.gapminder()
fig = px.line(gapminder, x="year", y="gdpPercap", color="country")
fig_copy = copy.deepcopy(fig)
```

The entire error trace gives some more insights, here what I could figure out (maybe it is trivial, yet it wasn't for me)

- `deepcopy` calls `__reduce__`, which calls  `to_dict`, which calls `convert_to_base64`.
- At this point every value is converted in the following form:
    ```py
    'x' = {'dtype': 'i2', 'bdata': 'oAelB6oHrwe0B7kHvgfDB8gHzQfSB9cH'},
    'y' = {'dtype': 'f8', 'bdata': 'ESBU8...7M'}
    ```
- Then `deepcopy` will call `_reconstruct`, which tries to run `Figure(*converted_values)`, hence `BaseFigure`, and finally this will call `validate_coerce` which will fail.

Hope it helps

File: packages/python/plotly/_plotly_utils/basevalidators.py (hunk 1 above)
Comment (marthacryan) [reply packages/python/plotly/_plotly_utils/basevalidators.py:1717]: Oh interesting! Yeah not sure why that's happening but it looks fine?

File: packages/python/plotly/_plotly_utils/tests/validators/test_fig_deepcopy.py [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+import plotly.express as px
+
+
+def test_deepcopy():
+    gapminder = px.data.gapminder()
+    fig = px.line(gapminder, x="year", y="gdpPercap", color="country")
+
+    fig_copy = copy.deepcopy(fig)
+
+    assert fig_copy is not None
```
Comment (marthacryan) [inline packages/python/plotly/_plotly_utils/tests/validators/test_fig_deepcopy.py:11]: This test looks good thank you for doing that! Do you think you could add a simple check that the data did copy correctly? Also, where are the other deepcopy tests you mentioned? Does it make more sense to include this test with those?

Comment (FBruzzesi) [reply packages/python/plotly/_plotly_utils/tests/validators/test_fig_deepcopy.py:11]: Thanks @marthacryan , what's the best way to assert that the values are the same? is it `fig_copy.to_dict() == fig.to_dict()`?

Here are few other figure deepcopy's in the test suite:
- [test_update_subplots](https://github.com/plotly/plotly.py/blob/725290068d5ea3f7f344eec7d7b84c58b85cfb38/packages/python/plotly/plotly/tests/test_core/test_update_objects/test_update_subplots.py#L364)
- [test_update_traces](https://github.com/plotly/plotly.py/blob/725290068d5ea3f7f344eec7d7b84c58b85cfb38/packages/python/plotly/plotly/tests/test_core/test_update_objects/test_update_traces.py#L295)
- [test_deepcopy_pickle.py](https://github.com/plotly/plotly.py/blob/725290068d5ea3f7f344eec7d7b84c58b85cfb38/packages/python/plotly/plotly/tests/test_io/test_deepcopy_pickle.py#L38)

File: packages/python/plotly/_plotly_utils/basevalidators.py (hunk 1 above)
Comment (FBruzzesi) [reply packages/python/plotly/_plotly_utils/basevalidators.py:1717]: Yeah I would not worry too much about this, I just wonder why that's the case 😇

**Review by marthacryan — APPROVED** [review]
Looks great, thanks @FBruzzesi!

File: packages/python/plotly/_plotly_utils/tests/validators/test_fig_deepcopy.py (hunk 2 above)
Comment (marthacryan) [reply packages/python/plotly/_plotly_utils/tests/validators/test_fig_deepcopy.py]: I think using `fig.to_dict()` is fine here. There are other ways to compare them but I just wanted a baseline check that the data is still there

**Pull request merged**